storage/remote/client.go

```diff
@@ -262,22 +271,40 @@ func (c *Client) Read(ctx context.Context, query *prompb.Query) (*prompb.QueryRe
 	if err != nil {
 		return nil, errors.Wrap(err, "error sending request")
 	}
+
+	contentType := httpResp.Header.Get("Content-Type")
+	if strings.HasPrefix(contentType, "application/x-protobuf") {
+		return c.handleSampledResponse(httpResp, start, req, sortSeries)
+	}
+
+	if httpResp.StatusCode/100 != 2 {
+		httpResp.Body.Close()
+		return nil, errors.Errorf("remote server %s returned HTTP status %s: %s", c.url.String(), httpResp.Status, strings.TrimSpace(string(compressed)))
+	}
+
+	if !strings.HasPrefix(contentType, "application/x-streamed-protobuf; proto=prometheus.ChunkedReadResponse") {
+		httpResp.Body.Close()
+		return nil, errors.Errorf("not supported remote read content type: %s", contentType)
+	}
+
+	// Handling streamed response
+	cancel = nil
+	stream := NewChunkedReader(httpResp.Body, DefaultChunkedReadLimit, nil)
+	return &chunkedResponseSeriesSet{stream: stream, httpResp: httpResp}, nil
+}
+
+func (c *Client) handleSampledResponse(httpResp *http.Response, start time.Time, req *prompb.ReadRequest, sortSeries bool) (storage.SeriesSet, error) {
 	defer func() {
 		io.Copy(ioutil.Discard, httpResp.Body)
 		httpResp.Body.Close()
 	}()
 	c.readQueriesDuration.Observe(time.Since(start).Seconds())
 	c.readQueriesTotal.WithLabelValues(strconv.Itoa(httpResp.StatusCode)).Inc()
 
-	compressed, err = ioutil.ReadAll(httpResp.Body)
+	compressed, err := ioutil.ReadAll(httpResp.Body)
 	if err != nil {
 		return nil, errors.Wrap(err, fmt.Sprintf("error reading response. HTTP status code: %s", httpResp.Status))
 	}
-
-	if httpResp.StatusCode/100 != 2 {
-		return nil, errors.Errorf("remote server %s returned HTTP status %s: %s", c.url.String(), httpResp.Status, strings.TrimSpace(string(compressed)))
-	}
-
```

Comment:
I'm not sure removing this is correct - for the sample response path, we never seen to check the status code?

Reply:
Fixed.

Reply:
How? We still have it removed (:

Reply:
I moved this code to `Client.Read` to handle sample/stream response path, like:
```go
	start := time.Now()
	httpResp, err := c.Client.Do(httpReq)
	if err != nil {
		return nil, errors.Wrap(err, "error sending request")
	}

	if httpResp.StatusCode/100 != 2 {
		httpResp.Body.Close()
		return nil, errors.Errorf("remote server %s returned HTTP status %s: %s", c.url.String(), httpResp.Status, strings.TrimSpace(string(compressed)))
	}

```